src/server/db_slice.cc

```diff
@@ -815,10 +811,7 @@ void DbSlice::FlushDbIndexes(const std::vector<DbIndex>& indexes) {
 }
 
 void DbSlice::FlushDb(DbIndex db_ind) {
-  // We should not flush if serialization of a big value is in progress because this
-  // could lead to UB or assertion failures (while DashTable::Traverse is iterating over
-  // a logical bucket).
-  util::fb2::LockGuard lk(local_mu_);
+  block_counter_.Wait();
```

Comment:
is this true? we can not flush while we are in the middle of a big value serialization? 
when snapshoting we can run flushdb as we use intrusive ptr there

Reply:
Well it did mess with `Traverse` in the past as it caused the assertion to trigger. Now we got bucket (not logical) Traversal it's probably ok. I would rather keep this though as a safe guard, and have another branch run this without this Wait here. We can incrementally remove those parts if you are ok with this

Reply:
I dont see how this is related to travesing on logical/ not logical bucket, we copy the pointer asside while flushing and snapshoting so that when flush is run we reset the pointer for the next transactions running.
Actually if we have a bug here I would like to understand before we merge this PR

Reply:
they should not be freed by flushall, what is the different if you preempt from snapshoting in the middle of entry serialize or at the end of bucket serialize.
when running flushall the snapshot should continue to create a point in time snapshot for the data although flush all finised and next transactions get empty db
this is why we use intrusive ptr in snapshoting

Reply:
```
148 // We use reference counting semantics of DbTable when doing snapshotting.
149 // There we need to preserve the copy of the table in case someone flushes it during
150 // the snapshot process. We copy the pointers in StartSnapshotInShard function.
151 using DbTableArray = std::vector<boost::intrusive_ptr<DbTable>>;  
```

and 

```
   db_array_ = slice->databases();
```

Now it makes sense.

Reply:
Something that pops my mind: 

Within FlushDbIndexes: 

```
   803                                                                                                   
   804   auto cb = [indexes, flush_db_arr = std::move(flush_db_arr)]() mutable {
   805     flush_db_arr.clear();
   806     ServerState::tlocal()->DecommitMemory(ServerState::kDataHeap | ServerState::kBackingHeap |
   807                                           ServerState::kGlibcmalloc);
   808   };
   809 
   810   fb2::Fiber("flush_dbs", std::move(cb)).Detach();

```

The fiber will clear `flush_db_arr` and then `call memory decommit`. This won't have any effect *if we are sharpshooing*. Shall we run `memory decommit` in the destructor instead ?

Reply:
yes we can

Reply:
nice. It's minor but I will patch this at some point.